File: common/src/main/java/juuxel/adorn/platform/NetworkBridge.java
package juuxel.adorn.platform;

import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.network.protocol.game.ClientGamePacketListener;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.network.protocol.Packet;
import net.minecraft.server.level.ServerLevel;

public interface NetworkBridge {
    void sendToTracking(Entity entity, Packet<? super ClientGamePacketListener> packet);
    void sendToClient(Player player, CustomPacketPayload payload);

    default void syncBlockEntity(BlockEntity be) {
        if (!(be.getLevel() instanceof ServerLevel world)) {
            throw new IllegalStateException("[Adorn] Block entities cannot be synced client->server");
        }
        world.getChunkSource().blockChanged(be.getBlockPos());
    }
}

File: common/src/main/kotlin/juuxel/adorn/platform/NetworkBridge.kt
package juuxel.adorn.platform

import juuxel.adorn.fluid.FluidReference
import net.minecraft.block.entity.BlockEntity
import net.minecraft.entity.Entity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.network.Packet
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.Identifier

interface NetworkBridge {
    fun sendToTracking(entity: Entity, packet: Packet<*>)
    fun createEntitySpawnPacket(entity: Entity): Packet<*>
    fun sendOpenBookPacket(player: PlayerEntity, bookId: Identifier)
    fun sendBrewerFluidSync(player: PlayerEntity, syncId: Int, fluid: FluidReference)

    fun syncBlockEntity(be: BlockEntity) {
        // Logic taken from Fabric API's BlockEntityClientSerializable.sync()
        val world = be.world
        if (world !is ServerWorld) throw IllegalStateException("[Adorn] Block entities cannot be synced client->server")
        world.chunkManager.markForUpdate(be.pos)
    }
}
